torch/nested/_internal/cached_tensor.py

```diff
@@ -0,0 +1,166 @@
+        self.nested_int_ref: Any = None
+
+    def __repr__(self) -> str:  # type: ignore[override]
+        source_repr = repr(
+            _get_source(self.metadata, self.source_fields, self.target_field)
+        )
+        return f"CachedTensor({source_repr})"
+
+    def __getattr__(self, name: str) -> torch.Tensor:
+        if name in self.metadata:
+            return self.metadata[name]
+        else:
+            raise AttributeError(
+                f"{type(self).__name__} object has no attribute '{name}'"
+            )
+
+    def __tensor_flatten__(self) -> Tuple[List[str], Dict[str, Any]]:
+        ctx = {
+            "source_fields": self.source_fields,
+            "extra_fields": self.extra_fields,
+        }
+        return [x for x in self.all_fields if self.metadata.get(x) is not None], ctx
+
+    @staticmethod
+    def __tensor_unflatten__(
+        inner_tensors: Dict, meta: Dict, outer_size: Any, outer_stride: Any
+    ) -> "CachedTensor":
+        return CachedTensor(inner_tensors, **meta)
+
+    @classmethod
+    def __torch_dispatch__(
+        cls, op: Any, types: Any, args: Any, kwargs: Any
+    ) -> torch.Tensor:
+        # Ensure that any registered ops are loaded
+        import torch.nested._internal.ops  # noqa: F401
+
+        # Doing any operation on a CachedTensor automatically unwraps and returns a non-CachedTensor
+        # We can improve this to do smarter things, like automatically cache .diff(), .cumsum(), etc.
+        if kwargs is None:
+            kwargs = {}
+
+        if op in _func_registry:
+            return _func_registry[op](op, *args, **kwargs)
+
+        unwrapped_args = pytree.tree_map_only(
+            CachedTensor,
+            lambda x: _get_source(x.metadata, x.source_fields, x.target_field),
+            args,
+        )
+        unwrapped_kwargs = pytree.tree_map_only(
+            CachedTensor,
+            lambda x: _get_source(x.metadata, x.source_fields, x.target_field),
+            kwargs,
+        )
+        return op(*unwrapped_args, **unwrapped_kwargs)
+
+
+torch.serialization.add_safe_globals([CachedTensor])
+
+
+_func_registry: Dict[Any, Callable] = {}
+
+
+@contextmanager
+def set_func_registry(registry: Dict[Any, Callable]) -> Generator:
+    global _func_registry
+    # Save the current registry to restore it later
+    old_registry = _func_registry
+    # Set the new registry
+    _func_registry = registry
+    try:
+        yield
+    finally:
+        # Restore the old registry
+        _func_registry = old_registry
+
+
+# Note: [ CacheTensor open registry ]
+#
+# Registering this decorator for an op allows CachedTensor's torch dispatch to
+# redirect calls to that op to your own implementations.
+# For NestedTensor we rely on this behavior for factory functions.
+def register_cached_tensor_func(aten_op: Any) -> Callable:
+    def wrapper(func: Callable) -> Callable:
+        _func_registry[aten_op] = func
+        return func
+
+    return wrapper
```

Comment:
I think it's an interesting way to customize the behavior of this class based on what the parent subclass needs. For the lazy: what in particular is this utilized for for NJT / torch.compile?

Reply:
Specifically, NJT has two constructors from_padded and from_jagged that rely on this. Both require CachedTensor as input.